contracts/utils/cryptography/RSA.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Math} from "../math/Math.sol";
+
+/**
+ * @dev RSA PKCS#1 v1.5 signature verification implementation according to https://datatracker.ietf.org/doc/html/rfc8017[RFC8017].
+ *
+ * This library supports PKCS#1 v1.5 padding to avoid malleability via chosen plaintext attacks in practical implementations.
+ * The padding follows the EMSA-PKCS1-v1_5-ENCODE encoding definition as per section 9.2 of the RFC. This padding makes
+ * RSA semanticaly secure for signing messages.
+ *
+ * Inspired by https://github.com/adria0/SolRsaVerify[Adrià Massanet's work]
+ */
+library RSA {
+    /**
+     * @dev Same as {pkcs1} but using SHA256 to calculate the digest of `data`.
+     */
+    function pkcs1Sha256(
+        bytes memory data,
+        bytes memory s,
+        bytes memory e,
+        bytes memory n
+    ) internal view returns (bool) {
+        return pkcs1(sha256(data), s, e, n);
+    }
+
+    /**
+     * @dev Verifies a PKCSv1.5 signature given a digest according the verification
+     * method described in https://datatracker.ietf.org/doc/html/rfc8017#section-8.2.2[section 8.2.2 of RFC8017].
+     *
+     * IMPORTANT: Although this function allows for it, using n of length 1024 bits is considered unsafe.
+     * Consider using at least 2048 bits.
+     *
+     * @param digest the digest to verify
+     * @param s is a buffer containing the signature
+     * @param e is the exponent of the public key
+     * @param n is the modulus of the public key
+     */
+    function pkcs1(bytes32 digest, bytes memory s, bytes memory e, bytes memory n) internal view returns (bool) {
+        unchecked {
+            // cache and check length
+            uint256 length = n.length;
+            if (
+                length < 0x40 || // PKCS#1 padding is slightly less than 0x40 bytes at the bare minimum
+                length != s.length // signature must have the same length as the finite field
+            ) {
+                return false;
+            }
+
+            // verify that s < n
+            bool ok = false;
+            for (uint256 i = 0; i < length; i += 0x20) {
+                bytes32 si = _unsafeReadBytes32(s, Math.min(i, length - 0x20));
+                bytes32 ni = _unsafeReadBytes32(n, Math.min(i, length - 0x20));
+                if (si < ni) {
+                    ok = true;
+                    break;
+                }
```

Comment:
Shall we mention this?

```suggestion
                // if (si > ni) return false; // Assuming an honest input, this check makes it more expensive for the caller
                if (si < ni) {
                    ok = true;
                    break;
                }
```

Reply:
We know `length >= 0x40` from a check above

Reply:
I don't understand the comment.

Reply:
If `si < ni`, then it's not needed to check the next memory word (32 bytes).  However, this only optimizes for an attacker, and wouldn't be needed for an honest caller.

I just found it weird that it loops until the end.

Reply:
actually we had a bug here